### PR TITLE
luac

### PR DESCRIPTION
**Pull request by giann**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,8 +6,7 @@
 	"directories": {
 		"lib": "src"
 	},
-	"scripts": {
-	},
+	"scripts": {},
```
Comment (daurnimator) [inline package.json]: unrelated change?

File: src/luac-cli.js [hunk 2]
```diff
@@ -0,0 +1,430 @@
+#!/usr/bin/env node
+"use strict";
+
+global.WEB = false;
+
+const lua      = require('../node_modules/fengari/src/lua.js');
```
Comment (daurnimator) [inline src/luac-cli.js]: Will these require paths work if someone does a 'global install' of this?

Comment (giann) [reply src/luac-cli.js]: Yes it works.
I also added the two binaries as `fengari` and `fengaric` in `package.json`.

You can test it with: `npm i -g fengari-lua/fengari-node-cli#luac`

Comment (daurnimator) [reply src/luac-cli.js]: I tried with:
```
sudo yarn global add fengari-lua/fengari-node-cli#luac
```

And it failed with:
```
$ fengaric
module.js:487
    throw err;
    ^

Error: Cannot find module '../node_modules/fengari/src/lua.js'
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.Module._load (module.js:437:25)
    at Module.require (module.js:513:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/usr/local/share/.config/yarn/global/node_modules/fengari-node-cli/src/luac-cli.js:6:18)
    at Module._compile (module.js:569:30)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:503:32)
    at tryModuleLoad (module.js:466:12)
    at Function.Module._load (module.js:458:3)
```

Comment (giann) [reply src/luac-cli.js]: Do you have any other fengari packages installed globally ?

Comment (daurnimator) [reply src/luac-cli.js]: Nope. That was the *only* node package I installed globally

Comment (giann) [reply src/luac-cli.js]: yarn install dependencies in a flat structure unlike npm it seems... not sure how to go around that

**Comment by daurnimator** [comment]
Slight differences to luac.

examples:
```
 2 params, 14 slots, 3 upvalues, 12 locals, 19 constants, 0 functions
```
vs
```
 2 params, 14 slots, 3 upvalues, 12 locals, 19 constants, 0 function
```

```
 	22	[92]	EQ       	0 11 -10	; - "Author"
```
vs
```
 	22	[92]	EQ       	0 11 -10	; -"Author"
```

File: src/luac-cli.js (hunk 2 above)
Comment (daurnimator) [reply src/luac-cli.js]: seems fixed now :)

**Comment by giann** [comment]
Should be the same now

**Comment by daurnimator** [comment]
Some differences:

```diff
-	46	[312]	DIV      	11 -5 11	; 1.0 -
+	46	[312]	DIV      	11 -5 11	; 1.0- 
```

```diff
-	20	[279]	SETTABLE 	8 -1 -1	; 1 1
+	20	[279]	SETTABLE 	8 -1 -1	; 11
```

```diff
-	12	[351]	CLOSURE  	4 0	; 0xa9ddb0
+	12	[351]	CLOSURE  	4 bx	; 0x39
```

```diff
-	22	[378]	MUL      	6 6 -2	; - 1e-06
+	22	[378]	MUL      	6 6 -2	; - 0.000001
```

File: src/luac-cli.js [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+const CT       = require(`${fengariPath}/defs.js`).CT;
+const lauxlib  = require(`${fengariPath}/lauxlib.js`);
+const luaconf  = require(`${fengariPath}/luaconf.js`);
+const lopcodes = require(`${fengariPath}/lopcodes.js`);
+const ops      = lopcodes.OpCodesI;
+
+const fs       = require("fs");
+const sprintf  = require("sprintf-js").sprintf;
+
+const PROGNAME = "luac";      /* default program name */
```
Comment (daurnimator) [inline src/luac-cli.js]: Should this be fengaric?

File: src/luac-cli.js [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+
+const doargs = function() {
+    let version = 0;
+    let i;
+    for (i = 2; i < process.argv.length; i++) {
+        let arg = process.argv[i];
+
+        if (arg[0] !== "-")  /* end of options; keep it */
+            break;
+
```
Comment (daurnimator) [inline src/luac-cli.js]: random newline?

File: src/luac-cli.js [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+        if (version === process.argv.length - 2) process.exit(0);
+    }
+    return i;
+};
+
+const FUNCTION = "(function()end)();";
+
+const reader = function(L, ud) {
+    if (ud--)
+        return lua.to_luastring(FUNCTION);
```
Comment (daurnimator) [inline src/luac-cli.js]: Could you move the `lua.to_luastring` to outside this function?

File: src/luac-cli.js [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    else
+        return null;
+};
+
+const combine = function(L, n) {
+    if (n === 1)
+        return lua.lua_topointer(L, -1).p;
```
Comment (daurnimator) [inline src/luac-cli.js]: Could we write a `toproto` helper function to match the C code better?

File: src/luac-cli.js [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+            if (f.p[i].upvalues.length > 0)
+                f.p[i].upvalues[0].instack = false;
+        }
+        f.lineinfo.length = 0;
+        return f;
+    }
+};
+
+const writer = function(L, p, size, u) {
+    return fs.writeSync(u, new Buffer(p), 0, size) > 0 ? 0 : 1;
```
Comment (daurnimator) [inline src/luac-cli.js]: Should this be `Buffer.from`?

File: src/luac-cli.js [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+    let argv = lua.lua_touserdata(L,2);
+
+    let i;
+    if (!lua.lua_checkstack(L,argc))
+        fatal("too many input files");
+
+    for (i = 0; i < argc; i++) {
+        let filename = argv[i] === "-" ? null : argv[i];
+        if (lauxlib.luaL_loadfile(L, lua.to_luastring(filename)) !== lua.LUA_OK)
+            fatal(lua.to_jstring(lua.lua_tostring(L,-1)));
```
Comment (daurnimator) [inline src/luac-cli.js]: Use `lua_tojsstring`?

File: src/luac-cli.js [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+        usage("no input files given");
+
+    let L = lauxlib.luaL_newstate();
+
+    lua.lua_pushcfunction(L, pmain);
+    lua.lua_pushinteger(L, process.argv.length - i);
+    lua.lua_pushlightuserdata(L, process.argv.slice(i));
+
+    if (lua.lua_pcall(L,2,0,0) !== lua.LUA_OK)
+        fatal(lua.lua_tostring(L,-1));
```
Comment (daurnimator) [inline src/luac-cli.js]: Should this be `lua_tojsstring`?

File: src/luac-cli.js [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+            case '\f': print("\\f"); break;
+            case '\n': print("\\n"); break;
+            case '\r': print("\\r"); break;
+            case '\t': print("\\t"); break;
+            case '\v': print("\\v"); break;
+            default:
+                if (isprint(c))
+                    print(String.fromCharCode(c));
+                else
+                    print(`0x${c.toString(16)}`);
```
Comment (daurnimator) [inline src/luac-cli.js]: luac uses decimal escapes, not hex ones.

File: src/luac-cli.js [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+                if (lopcodes.getBMode(o) != lopcodes.OpArgN)
+                    print(` ${lopcodes.ISK(b) ? MYK(lopcodes.INDEXK(b)) : b}`);
+                if (lopcodes.getCMode(o) != lopcodes.OpArgN)
+                    print(` ${lopcodes.ISK(c) ? MYK(lopcodes.INDEXK(c)) : c}`);
+                break;
+            }
+            case lopcodes.iABx: {
+                print(`${a}`);
+                if (lopcodes.getBMode(o) == lopcodes.OpArgK) print(` ${MYK(bx)}`);
+                if (lopcodes.getBMode(o) == lopcodes.OpArgU) print(` bx`);
```
Comment (daurnimator) [inline src/luac-cli.js]: missing interpolation

File: src/luac-cli.js [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+            case ops.OP_SHL:
+            case ops.OP_SHR:
+            case ops.OP_EQ:
+            case ops.OP_LT:
+            case ops.OP_LE: {
+                if (lopcodes.ISK(b) || lopcodes.ISK(c)) {
+                    print("\t; ");
+                    if (lopcodes.ISK(b)) PrintConstant(f, lopcodes.INDEXK(b));
+                    else print("- ");
+                    if (lopcodes.ISK(c)) PrintConstant(f, lopcodes.INDEXK(c));
```
Comment (daurnimator) [inline src/luac-cli.js]: Missing print of space before this line

File: src/luac-cli.js [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+            case ops.OP_BXOR:
+            case ops.OP_SHL:
+            case ops.OP_SHR:
+            case ops.OP_EQ:
+            case ops.OP_LT:
+            case ops.OP_LE: {
+                if (lopcodes.ISK(b) || lopcodes.ISK(c)) {
+                    print("\t; ");
+                    if (lopcodes.ISK(b)) PrintConstant(f, lopcodes.INDEXK(b));
+                    else print("- ");
```
Comment (daurnimator) [inline src/luac-cli.js]: This print shouldn't have a space after the hyphen

File: src/luac-cli.js [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+                    if (lopcodes.ISK(c)) PrintConstant(f, lopcodes.INDEXK(c));
+                    else print("- ");
```
Comment (daurnimator) [inline src/luac-cli.js]: This print shouldn't have a space after the hyphen

File: src/luac-cli.js [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+const SS = function(n) {
+    return n === 1 ? "" : "s";
+};
+
+const PrintHeader = function(f) {
+    let s = f.source ? f.source.getstr() : lua.to_luastring("=?");
+    if (s[0] === "@".charCodeAt(0) || s[0] === "=".charCodeAt(0))
+        s = s.slice(1);
+    else if (s[0] === lua.LUA_SIGNATURE.charCodeAt(0))
+        s = lua.to_luastring("(bstring");
```
Comment (daurnimator) [inline src/luac-cli.js]: missing close parenthesis

File: src/luac-cli.js [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+    else
+        s = lua.to_luastring("(string");
```
Comment (daurnimator) [inline src/luac-cli.js]: missing close parenthesis

File: src/luac-cli.js [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+
+    n = f.upvalues.length;
+    print(`upvalues (${n}) for 0x${f.id.toString(16)}:\n`);
+    for (let i = 0; i < n; i++) {
+        print(`\t${i}\t${UPVALNAME(f, i)}\t${f.upvalues[i].instack ? 1 : 0}\t${f.upvalues[i].idx}\n`);
+    }
+};
+
+const PrintFunction = function(f, full) {
+    PrintHeader(f);
```
Comment (daurnimator) [inline src/luac-cli.js]: Might as well use a var `n` like the C code

**Review by daurnimator — CHANGES_REQUESTED** [review]

File: src/luac-cli.js (hunk 8 above)
Comment (giann) [inline src/luac-cli.js]: Yes and `new Buffer` is even [deprecated](https://nodejs.org/api/buffer.html#buffer_new_buffer_array)

File: src/luac-cli.js (hunk 7 above)
Comment (giann) [reply src/luac-cli.js]: Yes and `new Buffer` is even [deprecated](https://nodejs.org/api/buffer.html#buffer_new_buffer_array)

**Review by daurnimator — APPROVED** [review]

**Comment by daurnimator** [comment]
Looks good now except for rendering of numbers; but that appears to be an issue with sprintf.js.

**Pull request merged**